### PR TITLE
test(borrow): add a test to ensure borrow works

### PR DESCRIPTION
**Pull request by Ryman**
Ran into this when trying to remove some unsafety from hyper. 

This works on stable but seems to be a regression on beta and nightly. Placing this here for easy reference/repro.

```rust
src/lib.rs:142:17: 142:68 error: cannot infer an appropriate lifetime due to conflicting requirements [E0495]
src/lib.rs:142             map.get::<UniCase<Cow<'key, str>>>(&UniCase(fn_borrow))
                               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
src/lib.rs:140:9: 143:10 help: consider using an explicit lifetime parameter as shown: fn get_shorter<'key>(map: &'key HashMap<UniCase<Cow<'static, str>>, u32>,
                     s: &'key str) -> Option<&'key u32>
src/lib.rs:140         fn get_shorter<'map, 'key>(map: &'map HashMap<UniCase<Cow<'static, str>>, u32>, s: &'key str) -> Option<&'map u32>{
src/lib.rs:141             let fn_borrow: Cow<'key, str> = s.into();
src/lib.rs:142             map.get::<UniCase<Cow<'key, str>>>(&UniCase(fn_borrow))
src/lib.rs:143         }
error: aborting due to previous error
```

Note that it works if you adjust the lifetimes to the suggestion, but that would limit the return value's lifetime to the lifetime of the key which would be undesirable for many cases.

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -108,4 +110,23 @@ mod test {
...
+
+        let mut map = HashMap::new();
+        let val = 42;
+        map.insert(UniCase(owned), val);
+
+        assert_eq!(map.get(&UniCase(static_borrow)), Some(&val));
+
+        fn get_shorter<'map, 'key>(map: &'map HashMap<UniCase<Cow<'static, str>>, u32>, s: &'key str) -> Option<&'map u32>{
+            let fn_borrow: Cow<'key, str> = s.into();
+            map.get::<UniCase<Cow<'key, str>>>(&UniCase(fn_borrow))
```
Comment (Ryman) [inline src/lib.rs]: This `get` works with and without the type-hint on stable (1.4), fails in both cases on beta and nightly.

**Comment by seanmonstar** [comment]
Is this testing something specific about UniCase, or just lifetimes and Cow?

**Comment by Ryman** [comment]
The main intention was to test that `Borrow` actually worked *through* unicase, ensuring it works with HashMaps, but you're right in that it's not a very specific test. The verdict on the rust issue seems to be that it's a soundness issue and it's unlikely to compile as-is in the future.

**Comment by Ryman** [comment]
Closing for now, don't think there's a good way to do this currently.

**Pull request closed without merging**